File: current_session/1448.py
"""
1448
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        # use dfs
        self.count = 0
        
        def dfs(node, maxSeen):
            if not node: return
            if node.val >= maxSeen:     # good
                maxSeen = node.val
                self.count += 1
            dfs(node.right, maxSeen)
            dfs(node.left, maxSeen)
        
        dfs(root, root.val)

        return self.count

File: current_session/c++/1448.cpp
#include <iostream>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
  int cnt = 0;
  int goodNodes(TreeNode* root) {
    dfs(root, INT_MIN);
    return cnt;
  }
  void dfs(TreeNode* node, int maxValInPath) {
    if (node == nullptr) return;
    if (node->val >= maxValInPath) cnt++;
    if (node->left != nullptr) dfs(node->left, max(maxValInPath, node->val));
    if (node->right != nullptr) dfs(node->right, max(maxValInPath, node->val));
    return;
  }
};